Support input/output in vector registers of s390x inline assembly (under asm_experimental_reg feature)

taiki-e opened the pull request:
This extends currently clobber-only vector registers (`vreg`) support to allow passing `#[repr(simd)]` types, floats (f32/f64/f128), and integers (i32/i64/i128) as input/output.

This is unstable and gated under new `#![feature(asm_experimental_reg)]` (tracking issue: https://github.com/rust-lang/rust/issues/133416). If the feature is not enabled, only clober is supported as before.

| Architecture | Register class | Target feature | Allowed types |
| ------------ | -------------- | -------------- | -------------- |
| s390x | `vreg` | `vector` | `i32`, `f32`, `i64`, `f64`, `i128`, `f128`, `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4`, `f64x2` |

This matches the list of types that are supported by the vector registers in LLVM:
https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0/llvm/lib/Target/SystemZ/SystemZRegisterInfo.td#L301-L313

In addition to `core::simd` types and floats listed above, custom `#[repr(simd)]` types of the same size and type are also allowed. All allowed types other than i32/f32/i64/f64/i128, and relevant target features are currently unstable.

Currently there is no SIMD type for s390x in `core::arch`, but this is tracked in https://github.com/rust-lang/rust/issues/130869.

cc https://github.com/rust-lang/rust/issues/130869 about vector facility support in s390x
cc https://github.com/rust-lang/rust/issues/125398 & https://github.com/rust-lang/rust/issues/116909 about f128 support in asm

@rustbot label +O-SystemZ +A-inline-assembly

bors commented:
:umbrella: The latest upstream changes (presumably #132152) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #133006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

taiki-e committed:
Support input/output in vector registers of s390x inline assembly

taiki-e commented:
Vector ABI support (https://github.com/rust-lang/rust/pull/131586) has been merged, and this is now ready for review.

cc @uweigand
r? @Amanieu 

@rustbot label -S-blocked +S-waiting-on-review

Amanieu commented:
Is there a way to feature-gate this functionality? Since s390x asm support is stable, every change here needs to go through an FCP since it will instantly be stable.

taiki-e commented:
> Is there a way to feature-gate this functionality?

Yes, I implemented it in another branch (https://github.com/taiki-e/rust/compare/s390x-asm-vreg-inout...s390x-asm-vreg-inout-unstable):

```
error[E0658]: this register class can only be used as a clobber in stable
  --> $DIR/bad-reg.rs:74:9
   |
LL |         asm!("", in("v0") v); // requires vector & asm_experimental_arch
   |         ^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #93335 <https://github.com/rust-lang/rust/issues/93335> for more information
   = help: add `#![feature(asm_experimental_arch)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
```

Here I'm using the existing asm_experimental_arch as a gate, but it might make more sense to add a dedicated gate such as asm_experimental_reg.

taiki-e commented:
Added asm_experimental_reg feature and moved s390x non-clobber-only vector register support under that feature.

```
error[E0658]: this register class can only be used as a clobber in stable
  --> $DIR/bad-reg.rs:74:9
   |
LL |         asm!("", in("v0") v); // requires vector & asm_experimental_reg
   |         ^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #133416 <https://github.com/rust-lang/rust/issues/133416> for more information
   = help: add `#![feature(asm_experimental_reg)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11995604170/job/33439082299) [(plain)](https://github.com/rust-lang/rust/commit/d2357fbc9b8aac85df7e00c4eed83964cb8522d4/checks/33439082299/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.840 Building wheels for collected packages: reuse
#13 2.841   Building wheel for reuse (pyproject.toml): started
#13 3.092   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.093   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.093   Stored in directory: /tmp/pip-ephem-wheel-cache-pe6xqzmj/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.096 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.489 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.490 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.019 Collecting virtualenv
#13 4.019 Collecting virtualenv
#13 4.059   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.114      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 59.7 MB/s eta 0:00:00
#13 4.169 Collecting platformdirs<5,>=3.9.1
#13 4.175   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.194 Collecting distlib<1,>=0.3.7
#13 4.206      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 83.2 MB/s eta 0:00:00
#13 4.241 Collecting filelock<4,>=3.12.2
#13 4.245   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.245   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.326 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.501 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      329664 kB
DirectMap2M:     7010304 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5678 files
tidy check
tidy error: /checkout/compiler/rustc_ast_lowering/messages.ftl: message `ast_lowering_unstable_may_unwind` appears before `ast_lowering_register_class_only_clobber_stable`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Nov 24 11:22:00 UTC 2024
  network time: Sun, 24 Nov 2024 11:22:00 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Amanieu commented on compiler/rustc_codegen_cranelift/src/inline_asm.rs:
```diff
@@ -462,8 +465,12 @@ impl<'tcx> InlineAssemblyGenerator<'_, 'tcx> {
-            let reg_size =
-                reg_class.supported_types(self.arch).iter().map(|(ty, _)| ty.size()).max().unwrap();
+            let reg_size = reg_class
+                .supported_types(self.arch, self.allow_experimental_reg)
```
At the codegen stage we can just use `true` here instead of `allowed_experimental_reg`. All the stability checking will have been done in priot stages.

taiki-e replied:
Changed rustc_codegen_* to just use true.

I also left a document about the cases in which it is fine to always pass true.

https://github.com/rust-lang/rust/blob/c024d8ccdfb462f4e99b54e6c8456fbc4d745752/compiler/rustc_target/src/asm/mod.rs#L608-L610

Amanieu commented on compiler/rustc_hir_analysis/src/check/intrinsicck.rs:
```diff
@@ -218,8 +218,9 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
         let reg_class = reg.reg_class();
-        let supported_tys = reg_class.supported_types(asm_arch);
+        let supported_tys = reg_class.supported_types(asm_arch, allow_experimental_reg);
         let Some((_, feature)) = supported_tys.iter().find(|&&(t, _)| t == asm_ty) else {
```
This also needs to handle the case where a given type is only supported with allow_experimental_reg and report an appropriate error message. This would be used in the future for SVE support for example since it adds more types to an existing register class.

taiki-e replied:
Done.
https://github.com/rust-lang/rust/blob/c024d8ccdfb462f4e99b54e6c8456fbc4d745752/tests/ui/feature-gates/feature-gate-asm_experimental_reg.stderr#L11-L19

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11996122573/job/33440224011) [(plain)](https://github.com/rust-lang/rust/commit/7bd6de8d0c4c5acd620b058d9b6e1c51adedaf9f/checks/33440224011/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.889 Building wheels for collected packages: reuse
#13 2.890   Building wheel for reuse (pyproject.toml): started
#13 3.136   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.137   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.137   Stored in directory: /tmp/pip-ephem-wheel-cache-2yky3q2k/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.140 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.543 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.543 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.068 Collecting virtualenv
#13 4.068 Collecting virtualenv
#13 4.106   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.166      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 54.5 MB/s eta 0:00:00
#13 4.221 Collecting filelock<4,>=3.12.2
#13 4.228   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.263 Collecting platformdirs<5,>=3.9.1
#13 4.267   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.284 Collecting distlib<1,>=0.3.7
#13 4.296      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 80.5 MB/s eta 0:00:00
#13 4.296      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 80.5 MB/s eta 0:00:00
#13 4.376 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.555 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      305088 kB
DirectMap2M:     8083456 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5678 files
tidy check
tidy error: /checkout/compiler/rustc_hir_analysis/messages.ftl: message `hir_analysis_unused_generic_parameter_ty_alias_help` appears before `hir_analysis_register_type_unstable`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Nov 24 12:40:18 UTC 2024
  network time: Sun, 24 Nov 2024 12:40:18 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

taiki-e committed:
Make s390x non-clobber-only vector register support unstable

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit c024d8ccdfb462f4e99b54e6c8456fbc4d745752 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu c024d8ccdfb462f4e99b54e6c8456fbc4d745752 -->
<!-- homu: {"type":"Approved","sha":"c024d8ccdfb462f4e99b54e6c8456fbc4d745752","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
